usb/usb_host_hid/hid_host.c

```diff
@@ -0,0 +1,1465 @@
+        ESP_ERROR_CHECK(usb_host_device_info(hid_device->dev_hdl, &dev_info));
+        // reallocate the ctrl xfer buffer for new length
+        ESP_LOGI(TAG, "Change HID ctrl xfer size from %d to %d",
+                 hid_device->ctrl_xfer->data_buffer_size,
+                 (int) (dev_info.bMaxPacketSize0 + req->wLength));
+
+        usb_host_transfer_free(hid_device->ctrl_xfer);
+        HID_RETURN_ON_ERROR( usb_host_transfer_alloc(USB_SETUP_PACKET_SIZE + req->wLength,
+                             0,
+                             &hid_device->ctrl_xfer),
+                             "Unable to allocate transfer buffer for EP0");
+    }
+
+    HID_RETURN_ON_ERROR( hid_device_try_lock(hid_device, DEFAULT_TIMEOUT_MS),
+                         "HID Device is busy by other task");
+
+    usb_setup_packet_t *setup = (usb_setup_packet_t *)ctrl_xfer->data_buffer;
+
+    setup->bmRequestType = USB_BM_REQUEST_TYPE_DIR_IN |
+                           USB_BM_REQUEST_TYPE_TYPE_STANDARD |
+                           USB_BM_REQUEST_TYPE_RECIP_INTERFACE;
+    setup->bRequest = req->bRequest;
+    setup->wValue = req->wValue;
+    setup->wIndex = req->wIndex;
+    setup->wLength = req->wLength;
+
+    HID_RETURN_ON_ERROR( hid_control_transfer(hid_device,
+                         ctrl_xfer,
+                         USB_SETUP_PACKET_SIZE + req->wLength,
+                         DEFAULT_TIMEOUT_MS),
+                         "Control transfer failure");
+
+    ctrl_xfer->actual_num_bytes -= USB_SETUP_PACKET_SIZE;
+
+    if (ctrl_xfer->actual_num_bytes <= req->wLength) {
+        memcpy(req->data, ctrl_xfer->data_buffer + USB_SETUP_PACKET_SIZE, ctrl_xfer->actual_num_bytes);
+    }
+
+    hid_device_unlock(hid_device);
+
+    return ESP_OK;
+}
+
+/**
+ * @brief HID Host Request Report Descriptor
+ *
+ * @param[in] hidh_iface      Pointer to HID Interface configuration structure
+ * @return esp_err_t
+ */
+esp_err_t hid_class_request_report_descriptor(hid_iface_t *iface)
+{
+    HID_RETURN_ON_INVALID_ARG(iface);
+
+    // Get Report Descritpor is possible only in Ready or Active state
+    HID_RETURN_ON_FALSE((HID_INTERFACE_STATE_READY == iface->state) ||
+                        (HID_INTERFACE_STATE_ACTIVE == iface->state),
+                        ESP_ERR_INVALID_STATE,
+                        "Unable to request report decriptor. Interface is not ready");
+
+    iface->report_desc = calloc(1, iface->report_desc_size);
+    HID_RETURN_ON_FALSE(iface->report_desc,
+                        ESP_ERR_NO_MEM,
+                        "Unable to allocate memory");
+
+    const hid_class_request_t get_desc = {
+        .bRequest = USB_B_REQUEST_GET_DESCRIPTOR,
+        .wValue = (HID_CLASS_DESCRIPTOR_TYPE_REPORT << 8),
+        .wIndex = iface->dev_params.iface_num,
+        .wLength = iface->report_desc_size,
+        .data = iface->report_desc
+    };
+
+    return usb_class_request_get_descriptor(iface->parent, &get_desc);
```

Comment:
If the return value is not ESP_OK, I think memory that is allocated above in the following line will leak:
iface->report_desc = calloc(1, iface->report_desc_size);

Reply:
Lifetime of iface->report_desc is equal to iface lifetime, right @roma-jam ? so we free this memory only after the device was detached/removed